src/engines/nwn2/inventory.cpp

```diff
@@ -83,6 +83,26 @@ Item *Inventory::getNextItemInInventory() {
 	return 0;
 }
 
+Item *Inventory::createItem(const Common::UString &blueprint, uint16 stackSize, const Common::UString &tag) {
+	if (blueprint.empty() || stackSize == 0)
+		return 0;
+
+	// Create the item
+	Item *item = new Item(blueprint, stackSize, tag);
+	if (item) {
+		// Was the tag set?
+		if (!item->getTag().empty()) {
+			// TODO: Check for a merge into a stack of similar items
+			uint16 slot = getFirstEmptySlot();
+			inventoryItems[slot] = item;
+		} else {
+			delete(item);
```

Comment:
No need for parens on delete.

As an alternative, you could make item a std::unique_ptr (or a Common::ScopedPtr, but I'd like to phase usage of ScopedPtr out in favour of unique_ptr) and return item.release() on success.

Reply:
Okay.

Reply:
I'm intentionally making the items so that they would be transferable between inventories. Items can have inventories of other items, so it could potentially get recursive. That's the reason I was avoiding ScopedPtr. Does that make sense or am I being too paranoid? I think eventually an item will also need to carry a back pointer to its inventory so that a delete object can be performed.

Reply:
You can still use a ScopedPtr/unique_ptr locally, and call the release() method, which returns the raw pointer and releases it from the managed pointer. This is just so that you can return from the function in the error case without having to worry about manually deleting the object.

Reply:
If I'm understanding the std::unique_ptr functionality correctly, I can't return the object itself with the getFirstItemInInventory & getNextItemInInventory calls(?).

Reply:
You can.

```cpp
Foo *createFoo() [
  std::unique_ptr<Foo> foo(new Foo(23, 5);

  ...
  if (somethingFailed)
    return nullptr;
  ...

  return foo.release();
}

Reply:
Or maybe I don't understand your question?

Reply:
Well I don't think I want to release the ownership. The pointer is being accessed by script calls via the getFirstInInventory & getNextInInventory. But it should still remain in the inventory.

Reply:
No, that releases the pointer from the temporary local unique_ptr. The purpose of that local unique_ptr would be solely to manage the memory inside the function. With the return release, you transfer the ownership to the caller.

Reply:
We can talk about that in-depth on IRC if you want :)
Later today, though, because it's still early here and I'd like to try for some more sleep :P

Reply:
I'm thinking a std::shared_ptr might work better in this instance.

Reply:
Maybe. But that change is more intricate, so maybe that should be left for the future. While just swapping out the raw pointer within the function for a unique_ptr would be easy to do now.

Reply:
For example, I've got an item in a crafting table inventory. The PC runs a script that modifies the itemProperty list on the item, then exits, leaving the item still in the table inventory. The script will need to grab the item instance, make some changes, then exit.

Reply:
Yeah, I guess I'm not understanding how removing the unique_ptr from the inventory to dispatch it to a temporary script is helping. Oh well.

Reply:
Huh? Swapping in the std::unique_ptr for the raw pointer here changing nothing outside of the function. You're still returning the raw pointer. The release is only there to release it from the temporary unique_ptr that was created for the function.

Literally the only thing that changes is that you won't have to manually delete the item when an error in the middle of the function happens. That's the only purpose of the local unique_ptr. The benefit is that you can't forgot to delete it, so there's never a leak.

Reply:
I.e. this is what I'm talking about, rewriting this function to use a unique_ptr (and keeping the rest as-if, like the map, for clarity):

```cpp
Item *Inventory::createItem(const Common::UString &blueprint, uint16 stackSize, const Common::UString &tag) {
	if (blueprint.empty() || stackSize == 0)
		return nullptr;

	std::unique_ptr<Item> item(new Item(blueprint, stackSize, tag));
	if (item->getTag().empty())
		return nullptr;

	const uint16 slot = getFirstEmptySlot();
	inventoryItems[slot] = item.release();

	return inventoryItems[slot];
}
```

The unique_ptr makes sure that the object is always deleted correctly, unless release was called (which transferes ownership).

Reply:
Guess I'm confused then. According to the docs., calling release replaces the stored pointer in the unique_ptr with null. The inventory no longer has access to that item instance. That's not what I want to do.

Reply:
Yes, it replaces it with null *and returns the previous pointer*, so that it can be used somewhere else. Like, to store it inside the map here.

That's what `inventoryItems[slot] = item.release();` does, it transfers the ownership from the unique_ptr into the inventory map.

That local unique_ptr is there so that, effictively, the function owns the item, until it's released in the inventory map.

Reply:
Ah, okay. I was thinking at a completely different level, believing you wanted the vector element to be a unique_ptr. Gotcha. No you just want it there long enough to assign to the slot. Sure, no problem.

Reply:
Yes, exactly. Sorry for the confusion :)

Reply:
Thank you for helping me work through it. :)